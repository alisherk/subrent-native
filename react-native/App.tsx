import React, { useState } from 'react';
import { SafeAreaView } from 'react-native';
import { rootReducer } from './redux/reducers';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { RootNavigator } from './navigation/RootNavigator';
import { AppLoading } from 'expo';
import { fixFirebaseTimer } from './timerFix';
import * as Font from 'expo-font';
import thunk from 'redux-thunk';


fixFirebaseTimer(); 

let composedEnhancer: any;
if (process.env.NODE_ENV !== 'production') {
  composedEnhancer = composeWithDevTools(applyMiddleware(thunk));
} else {
  composedEnhancer = applyMiddleware(thunk);
}

const store = createStore(rootReducer, composedEnhancer);

const fetchFonts = () => {
  return Font.loadAsync({
    Roboto: require('native-base/Fonts/Roboto.ttf'),
    Roboto_medium: require('native-base/Fonts/Roboto_medium.ttf'),
  });
};

export default function App() {
  const [fontLoaded, setFontLoaded] = useState(false);
  if (!fontLoaded) {
    return (
      <AppLoading
        startAsync={fetchFonts}
        onFinish={() => setFontLoaded(true)}
      />
    );
  }

  return (
    <Provider store={store}>
        <SafeAreaView style={{ flex: 1 }}>
          <RootNavigator />
        </SafeAreaView>
    </Provider>
  );
}
