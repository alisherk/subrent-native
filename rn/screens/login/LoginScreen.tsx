import React, { useState, useEffect } from 'react';
import { StyleSheet, ViewStyle, View, TouchableNativeFeedback } from 'react-native';
import { Button, Icon } from 'react-native-elements';
import { Form } from 'components/form';
import { LoginFields } from './LoginFields';
import { SignUpFields } from './SignUpFields';
import { ResetPasswordScreen } from './ResetPasswordScreen';
import { LoginScreenProps } from 'navigation';
import { useDispatch } from 'react-redux';
import * as actions from 'redux/actions';
import { FontAwesome } from '@expo/vector-icons';

enum Screens {
  SIGNUP = 'Sign up',
  RESET_PASSWORD = 'Reset password',
  LOGIN = 'Login',
}

interface FormValues {
  email: string;
  password: string;
  username: string;
}

const RenderInputByScreen = ({
  screen,
  onPress,
}: {
  screen: Screens;
  onPress?: () => void;
}): JSX.Element => {
  if (screen === Screens.SIGNUP) {
    return <SignUpFields />;
  }
  if (screen === Screens.RESET_PASSWORD) {
    return <ResetPasswordScreen />;
  }
  return <LoginFields onPress={onPress} />;
};

export const LoginScreen = ({ navigation, route }: LoginScreenProps) => {
  const dispatch = useDispatch();
  const [screen, setScreen] = useState(Screens.LOGIN);

  const originRoute = route.params?.origin || 'Shop';

  const handleSwitchScreen = (passedScreen?: Screens): void => {
    //navigate to desired screen passed as an argument
    let starterScreen = Screens.LOGIN;
    if (passedScreen) {
      starterScreen = passedScreen;
      setScreen(starterScreen);
      navigation.setOptions({ title: starterScreen });
      return;
    }
    //when no argument exists, switch the screen based on current screen state
    if (screen === Screens.LOGIN) {
      starterScreen = Screens.SIGNUP;
      setScreen(starterScreen);
      navigation.setOptions({ title: starterScreen });
      return;
    }
    setScreen(starterScreen);
    navigation.setOptions({ title: starterScreen });
  };

  useEffect(() => {
    navigation.setOptions({
      headerLeft: () => (
        <Button
          icon={<Icon name='arrow-back' color='white' />}
          onPress={() => {
            navigation.navigate(originRoute);
            handleSwitchScreen(Screens.LOGIN);
          }}
        />
      ),
    });
  }, [navigation]);

  const handleSubmit = async (formValues: FormValues): Promise<void> => {
    try {
      if (screen === Screens.RESET_PASSWORD) {
        await dispatch(actions.sendPasswordReset(formValues.email));
        return;
      }
      if (screen === Screens.SIGNUP) {
        await dispatch(actions.registerUser(formValues));
      } else {
        await dispatch(actions.loginUser(formValues));  
      } 
    } catch (err) {}
  };

  const handleLoginWithGoogle = async (): Promise<void> => {
    try {
      await dispatch(actions.loginWithGoogle());
    } catch (err) {}
  };

  return (
    <Form>
      {(formState) => (
        <>
          <RenderInputByScreen
            screen={screen}
            onPress={() => handleSwitchScreen(Screens.RESET_PASSWORD)}
          />
          <View style={styles.btnContainer}>
            <Form.Button
              title='Submit'
              disabled={!formState?.isValid || formState?.isSubmitting}
              onSubmit={handleSubmit}
            />
            <Form.Button
              title={screen === Screens.LOGIN ? 'Sign up' : 'Go back'}
              onPress={() => handleSwitchScreen()}
              type='outline'
            />
          </View>
          <Form.Button
            handlePress={handleLoginWithGoogle}
            type='outline'
            containerStyle={{ alignItems: 'center', padding: 0}}
            buttonStyle={{ width: 80, borderRadius: 100, height: 80}}
            icon={<FontAwesome name='google' size={40} color='crimson' />}
          />
        </>
      )}
    </Form>
  );
};

interface Styles {
  googleBtnContainer: ViewStyle;
  btnContainer: ViewStyle;
}

const styles = StyleSheet.create<Styles>({
  btnContainer: {
    justifyContent: 'space-evenly',
    flexDirection: 'row',
  },
  googleBtnContainer: {
    justifyContent: 'center',
    marginTop: 10,
  },
});
