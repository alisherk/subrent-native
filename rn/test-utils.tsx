import React from 'react';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import { createStore, applyMiddleware } from 'redux';
import { rootReducer } from './redux/reducers';
import { PaginateHookData } from './hooks';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

//mock redux store cofig for dispathing action 
const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
export const returnWithStore = (args?: any) =>  mockStore(args);

//testing libriary config for returning render with redux
export function renderWithRedux(
    component: React.ReactNode,
    initialState?: any,
    {
      store = createStore(rootReducer, initialState, applyMiddleware(thunk)),
    } = {}
  ) {
    return {
      ...render(<Provider store={store}>{component}</Provider>),
      store,
    };
  }
  
  //function for passing props to components with react navigation
  export const createTestProps =<T extends {}> (props: T) => ({
    navigation: {
      navigate: jest.fn(),
      setOptions: jest.fn(), 
      addListener: jest.fn()
    },
    route: {
      params: {}
    },
    ...props,
  });

//mock data used in tests
export const mockRental = {
  id: '1',
  name: 'test',
  full_day_price: '10',
  half_day_price: '5',
  category: 'test',
  region: 'test',
  description: 'test',
  ownerImage: 'test',
  displayName: 'test',
  delivery: 'test',
  image: 'test',
  instructions: 'test', 
  confirmation_required: 'test',
  ownerUid: 'e9e9e9er', 
  expoToken: 'ajdjdudnfififh',
  g: { geopoint: { U: 1, k: 2 }}
};

export const mockPaginateHookReturn: PaginateHookData = {
  rentals: [mockRental], 
  loadMore: jest.fn, 
  resetLoad: jest.fn, 
  loading: false, 
  reloading: false, 
  error: null, 
  hasMore: true, 
}


