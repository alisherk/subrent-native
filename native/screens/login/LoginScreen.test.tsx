import React from 'react';
import { LoginScreen } from './LoginScreen';
import { renderWithRedux, createTestProps } from 'test-utils';
import { cleanup, fireEvent, waitFor } from '@testing-library/react-native';

describe('LoginScreen', () => {
  afterEach(cleanup);
  const props: any = createTestProps({ navigation: { setOptions: jest.fn()}});

  it('switches the screen', async () => {
    const screen = renderWithRedux(<LoginScreen {...props} />);
    fireEvent.press(screen.getByText(/Sign up/));
    await waitFor(() => {
      expect(screen.getByPlaceholderText(/Username/)).toBeTruthy();
      expect(screen.getByText(/Go back/)).toBeTruthy();
    });
  });

  it('renders error message on invalid email', async () => {
    const screen = renderWithRedux(<LoginScreen {...props} />);
    const input = screen.getByPlaceholderText(/Email/);
    fireEvent.changeText(input, 'test');
    await waitFor(() =>
      expect(screen.getByText(/Valid email is required/)).toBeTruthy()
    ); 
  });

  it('switches to reset screen', async () => {
    const screen = renderWithRedux(<LoginScreen {...props} />);
    const button = screen.getByText(/Reset/);
    fireEvent.press(button); 
    expect(screen.getByPlaceholderText(/Email/)).toBeTruthy();
    expect(screen.queryByText(/Username/)).toBeNull();
    expect(screen.queryByText(/Password/)).toBeNull();
  });

  it('renders without crashing', async () => {
    const { toJSON } = renderWithRedux(<LoginScreen {...props} />);
    await waitFor(() => Promise.resolve());
    expect(toJSON()).toMatchSnapshot();
  });
});
