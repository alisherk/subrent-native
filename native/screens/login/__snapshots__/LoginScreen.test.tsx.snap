// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen renders without crashing 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    automaticallyAdjustContentInsets={false}
    contentContainerStyle={
      Array [
        Object {
          "padding": 10,
        },
        undefined,
      ]
    }
    contentInset={
      Object {
        "bottom": 0,
      }
    }
    enableAutomaticScroll={true}
    enableOnAndroid={false}
    enableResetScrollToCoords={true}
    extraHeight={75}
    extraScrollHeight={0}
    getScrollResponder={[Function]}
    handleOnScroll={[Function]}
    keyboardDismissMode="interactive"
    keyboardOpeningTime={250}
    keyboardShouldPersistTaps="handled"
    keyboardSpace={0}
    onScroll={[Function]}
    padder={true}
    resetKeyboardSpace={[Function]}
    resetScrollToCoords={
      Object {
        "x": 0,
        "y": 0,
      }
    }
    scrollEventThrottle={1}
    scrollForExtraHeightOnAndroid={[Function]}
    scrollIntoView={[Function]}
    scrollToEnd={[Function]}
    scrollToFocusedInput={[Function]}
    scrollToPosition={[Function]}
    showsVerticalScrollIndicator={true}
    style={
      Object {
        "backgroundColor": "transparent",
        "flex": 1,
      }
    }
    update={[Function]}
    viewIsInsideTabBar={false}
  >
    <View>
      <View
        style={Object {}}
      >
        <View
          style={
            Array [
              Object {},
              Object {
                "height": 80,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={false}
            forwardedRef={[Function]}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 1,
                "borderColor": "#D9D5DC",
                "borderLeftWidth": 0,
                "borderRadius": undefined,
                "borderRightWidth": 0,
                "borderTopWidth": 0,
                "borderWidth": undefined,
                "flexDirection": "row",
                "marginLeft": 2,
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "left": 0,
                  "opacity": 1,
                  "paddingBottom": undefined,
                  "paddingTop": undefined,
                  "position": "absolute",
                  "right": 0,
                  "top": 18,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#575757",
                    "fontSize": 17,
                    "paddingRight": 5,
                    "paddingTop": 5,
                  }
                }
              />
            </View>
            <TextInput
              allowFontScaling={true}
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="Email"
              placeholderTextColor="#575757"
              rejectResponderTermination={true}
              style={
                Object {
                  "color": "#000",
                  "flex": 1,
                  "fontSize": 17,
                  "height": 50,
                  "paddingBottom": 7,
                  "paddingLeft": 5,
                  "paddingRight": 5,
                  "paddingTop": 3,
                  "top": 8,
                }
              }
              underlineColorAndroid="rgba(0,0,0,0)"
              value=""
            />
          </View>
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "marginLeft": 5,
              }
            }
          />
        </View>
        <View
          style={
            Array [
              Object {},
              Object {
                "height": 80,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={false}
            forwardedRef={[Function]}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderBottomWidth": 1,
                "borderColor": "#D9D5DC",
                "borderLeftWidth": 0,
                "borderRadius": undefined,
                "borderRightWidth": 0,
                "borderTopWidth": 0,
                "borderWidth": undefined,
                "flexDirection": "row",
                "marginLeft": 2,
                "opacity": 1,
              }
            }
          >
            <View
              style={
                Object {
                  "left": 0,
                  "opacity": 1,
                  "paddingBottom": undefined,
                  "paddingTop": undefined,
                  "position": "absolute",
                  "right": 0,
                  "top": 18,
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "#575757",
                    "fontSize": 17,
                    "paddingRight": 5,
                    "paddingTop": 5,
                  }
                }
              />
            </View>
            <TextInput
              allowFontScaling={true}
              editable={true}
              onBlur={[Function]}
              onChangeText={[Function]}
              onFocus={[Function]}
              placeholder="Password"
              placeholderTextColor="#575757"
              rejectResponderTermination={true}
              secureTextEntry={true}
              style={
                Object {
                  "color": "#000",
                  "flex": 1,
                  "fontSize": 17,
                  "height": 50,
                  "paddingBottom": 7,
                  "paddingLeft": 5,
                  "paddingRight": 5,
                  "paddingTop": 3,
                  "top": 8,
                }
              }
              underlineColorAndroid="rgba(0,0,0,0)"
              value=""
            />
          </View>
          <View
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "marginLeft": 5,
              }
            }
          />
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-evenly",
            }
          }
        >
          <View
            style={
              Array [
                Object {},
                Object {
                  "alignItems": "center",
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              forwardedRef={[Function]}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "alignSelf": "flex-start",
                  "backgroundColor": "#b5b5b5",
                  "borderBottomWidth": null,
                  "borderColor": "#007aff",
                  "borderLeftWidth": null,
                  "borderRadius": 5,
                  "borderRightWidth": null,
                  "borderTopWidth": null,
                  "borderWidth": undefined,
                  "elevation": 2,
                  "flexDirection": "row",
                  "height": 45,
                  "justifyContent": "space-between",
                  "marginVertical": 5,
                  "opacity": 1,
                  "padding": 10,
                  "paddingBottom": 6,
                  "paddingTop": 6,
                  "shadowColor": undefined,
                  "shadowOffset": undefined,
                  "shadowOpacity": undefined,
                  "shadowRadius": undefined,
                }
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#fff",
                    "fontFamily": "System",
                    "fontSize": 16.5,
                    "marginLeft": 0,
                    "marginRight": 0,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                  }
                }
              >
                Submit
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {},
                Object {
                  "alignItems": "center",
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              forwardedRef={[Function]}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "alignSelf": "flex-start",
                  "backgroundColor": "#007aff",
                  "borderBottomWidth": null,
                  "borderColor": "#007aff",
                  "borderLeftWidth": null,
                  "borderRadius": 5,
                  "borderRightWidth": null,
                  "borderTopWidth": null,
                  "borderWidth": undefined,
                  "elevation": 2,
                  "flexDirection": "row",
                  "height": 45,
                  "justifyContent": "space-between",
                  "marginVertical": 5,
                  "opacity": 1,
                  "padding": 10,
                  "paddingBottom": 6,
                  "paddingTop": 6,
                  "shadowColor": undefined,
                  "shadowOffset": undefined,
                  "shadowOpacity": undefined,
                  "shadowRadius": undefined,
                }
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#fff",
                    "fontFamily": "System",
                    "fontSize": 16.5,
                    "marginLeft": 0,
                    "marginRight": 0,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                  }
                }
              >
                Sign up
              </Text>
            </View>
          </View>
          <View
            style={
              Array [
                Object {},
                Object {
                  "alignItems": "center",
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              forwardedRef={[Function]}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "alignSelf": "flex-start",
                  "backgroundColor": "#007aff",
                  "borderBottomWidth": null,
                  "borderColor": "#007aff",
                  "borderLeftWidth": null,
                  "borderRadius": 5,
                  "borderRightWidth": null,
                  "borderTopWidth": null,
                  "borderWidth": undefined,
                  "elevation": 2,
                  "flexDirection": "row",
                  "height": 45,
                  "justifyContent": "space-between",
                  "marginVertical": 5,
                  "opacity": 1,
                  "padding": 10,
                  "paddingBottom": 6,
                  "paddingTop": 6,
                  "shadowColor": undefined,
                  "shadowOffset": undefined,
                  "shadowOpacity": undefined,
                  "shadowRadius": undefined,
                }
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#fff",
                    "fontFamily": "System",
                    "fontSize": 16.5,
                    "marginLeft": 0,
                    "marginRight": 0,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                  }
                }
              >
                Reset
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "center",
              "marginTop": 10,
            }
          }
        >
          <View
            style={
              Array [
                Object {},
                Object {
                  "alignItems": "center",
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={true}
              forwardedRef={[Function]}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "alignItems": "center",
                  "alignSelf": "flex-start",
                  "backgroundColor": "#d9534f",
                  "borderBottomWidth": null,
                  "borderColor": "#007aff",
                  "borderLeftWidth": null,
                  "borderRadius": 5,
                  "borderRightWidth": null,
                  "borderTopWidth": null,
                  "borderWidth": undefined,
                  "elevation": 2,
                  "flexDirection": "row",
                  "height": 45,
                  "justifyContent": "space-between",
                  "marginVertical": 5,
                  "opacity": 1,
                  "padding": 10,
                  "paddingBottom": 6,
                  "paddingTop": 6,
                  "shadowColor": undefined,
                  "shadowOffset": undefined,
                  "shadowOpacity": undefined,
                  "shadowRadius": undefined,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": undefined,
                      "fontSize": 12,
                    },
                    Array [
                      Object {
                        "color": "#000",
                        "fontSize": 30,
                      },
                      Object {
                        "color": "#fff",
                        "fontSize": 24,
                        "marginLeft": 16,
                        "marginRight": 16,
                        "paddingTop": 2,
                      },
                    ],
                    Object {
                      "fontFamily": "ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#fff",
                    "fontFamily": "System",
                    "fontSize": 16.5,
                    "marginLeft": 0,
                    "marginRight": 0,
                    "paddingLeft": 16,
                    "paddingRight": 16,
                  }
                }
              >
                Login with Google
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
